src/robot.coffee

```diff
@@ -184,24 +190,112 @@ class Robot
       ((msg) -> msg.message = msg.message.message; callback msg)
     )
 
+  # Public: Registers new middleware for execution after matching but before
+  # Listener callbacks
+  #
+  # middleware - A function that determines whether or not a given matching
+  #              Listener should be executed. The function is called with
+  #              (robot, listener, response, next, done). If execution should
+  #              continue (next middleware, Listener callback), the middleware
+  #              should call the 'next' function with 'done' as an argument.
+  #              If not, the middleware should call the 'done' function with
+  #              no arguments.
+  #
+  # Returns nothing.
+  listenerMiddleware: (middleware) ->
+    if middleware.length != 5
+      throw new Error("Incorrect number of arguments for middleware callback (expected 5, got #{middleware.length})")
+    @middleware.listener.push (robot, context, next, done) ->
+      middleware.call(undefined, robot, context.listener, context.response, next, done)
+    return undefined
+
+  # Execute all middleware in order and call 'next' with the latest 'done'
```

Comment:
Let's mark this as `Private`. I don't think it's something a script writer would need to be using at all.